src/main/java/frc/robot/commands/autonomous/TrajectoryCommandGroup.java

```diff
@@ -0,0 +1,188 @@
+    ListIterator<Command> executing = m_executingCommands.listIterator();
+
+    // Execute the previously triggered commands
+    while (executing.hasNext()) {
+      Command command = executing.next();
+
+      command.execute();
+
+      // Check if the command is finished
+      if (command.isFinished()) {
+        // End the command, which finished naturally if isFinished() returns true
+        command.end(false);
+
+        // Remove the command from the currently executing list since it's finished executing
+        executing.remove();
+      }
+    }
+
+    // Get the robot's current position on the field as a Translation2d
+    Translation2d currentPosition = m_drive.getPose().getTranslation();
+
+    // Check if any commands should be triggered by the robot's position
+    for (Map.Entry<Command, PositionTrigger> poseCommand : m_untriggeredCommands.entrySet()) {
+      Command command = poseCommand.getKey();
+      PositionTrigger goal = poseCommand.getValue();
+
+      // Execute the command if it just got triggered, or has been previously
+      if (currentPosition.getDistance(goal.position) <= goal.tolerance) {
+        // Initialize the command
+        command.initialize();
+
+        // Add the command to the currently executing list
+        m_executingCommands.add(command);
+      }
+    }
+
+    // Remove all executing commands from the untriggered map
+    m_executingCommands.stream().forEach(m_untriggeredCommands::remove);
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    // End the trajectory command
+    m_trajectoryCommand.end(interrupted);
+
+    // End all other commands that were executing when this command finished/was interrupted
+    for (Command command : m_executingCommands) {
+      command.end(interrupted);
+    }
+  }
+
+  @Override
+  public boolean isFinished() {
+    // Finishes when the robot has finished driving along the trajectory
+    return m_trajectoryCommand.isFinished();
+  }
+
+  /**
+   * Don't use this method to add commands; use {@link #addCommands(Map)} instead.
+   *
+   * @throws IllegalArgumentException Whenever this overload is called
+   */
+  @Override
+  public void addCommands(Command... commands) {
+    throw new IllegalArgumentException(
+        "Commands can't be added to a TrajectoryCommandGroup without position triggers.");
+  }
+
+  /**
+   * Adds the commands to this command group. They will be triggered near their supplied position
+   * triggers.
+   *
+   * @param commands The map of commands to add to this group.
+   */
+  public void addCommands(Map<Command, PositionTrigger> commands) {
+    // Commands shouldn't be used elsewhere
+    requireUngrouped(commands.keySet());
+
+    for (Map.Entry<Command, PositionTrigger> poseCommand : commands.entrySet()) {
+      Command command = poseCommand.getKey();
+
+      // Throw an exception if two commands share a subsystem dependency
+      if (!Collections.disjoint(m_requirements, command.getRequirements())) {
```

Comment:
Say we need to turn off and turn on our shooter multiple times along the path, is this possible right now? (since both commands require the same subsystem)

Reply:
That is possible if you wrap all of the commands in proxy schedule commands, since they don't inherit the dependencies of their inner commands. 

The reason I did it this way was because there isn't a way to programmatically verify when (or rather where) a command will end, so I figured I'd leave it to the programmer to reason through. Also, if two commands end up trying to run at the same time, the command scheduler will handle the command interruption logic, rather than that having to be handled within TrajectoryCommandGroup.

Reply:
I see you have a nice exception message. Sorry!